### PR TITLE
143322/143456 implement ukhoweeknumber array fields for newcell

### PR DESCRIPTION
**Pull request by harshda15125**
- Updated Sap Action configuration and models for new fields.
- Added XML field mapping for ukhoWeekNumber (WeekNo, CurrentAlphaCorrection & ValidFrom)for New ENC cell
- Added interface for ValidFrom field logic
- Added unit test cases 
- Improved Stryker score
- Updated functional test cases

### CODE REVIEW

File: src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs [hunk 1]
```diff
@@ -3,15 +3,17 @@
...
 using UKHO.ERPFacade.Common.Logging;
 using UKHO.ERPFacade.Common.Models;
+using UKHO.ERPFacade.Common.Providers;
 
 namespace UKHO.ERPFacade.API.Helpers
 {
     public class EncContentSapMessageBuilder : IEncContentSapMessageBuilder
     {
-        private readonly ILogger<EncContentSapMessageBuilder> _logger;
+        private static ILogger<EncContentSapMessageBuilder> s_logger;
```
Comment (pravinvarade10801) [inline src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs:12]: Can we keep this object variable as a readonly?

File: src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs [hunk 2]
```diff
@@ -3,15 +3,17 @@
...
         private readonly IXmlHelper _xmlHelper;
         private readonly IFileSystemHelper _fileSystemHelper;
         private readonly IOptions<SapActionConfiguration> _sapActionConfig;
+        private static IWeekDetailsProvider s_weekDetailsProvider;
```
Comment (pravinvarade10801) [inline src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs:16]: Can we make this object as readonly?

File: src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs [hunk 3]
```diff
@@ -434,5 +485,23 @@ private UnitOfSale GetUnitOfSaleForEncCell(List<UnitOfSale> listOfUnitOfSales, P
             }
             return unitOfSale!;
         }
+
+        private static string GetUkhoWeekNumberData(UkhoWeekNumber ukhoWeekNumber)
+        {
+            var validWeek = ukhoWeekNumber.Week.ToString("D2");
+            var concatedString = string.Join("", ukhoWeekNumber.Year, validWeek);
```
Comment (pravinvarade10801) [inline src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs:492]: please add meaningful name to variable

**Review by pravinvarade10801 — COMMENTED** [review]
Have some minor comments.

File: src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs [hunk 4]
```diff
@@ -339,12 +349,53 @@ private static XmlElement BuildAction(XmlDocument soapXml, Product product, Unit
...
+                                itemSubNode.InnerText = GetXmlNodeValue(thursdayDate);
+                                break;
+
+                            case WeekNo:
+                                string weekData = GetUkhoWeekNumberData(ukhoWeekNumber);
+                                itemSubNode.InnerText = GetXmlNodeValue(weekData);
+                                break;
+
+                            case Correction:
+                                itemSubNode.InnerText = ukhoWeekNumber.CurrentWeekAlphaCorrection ? GetXmlNodeValue("Y") : GetXmlNodeValue("N");
```
Comment (pravinvarade10801) [inline src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs:375]: Can we add constant variable for both Correction values

File: src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs (hunk 3 above)
Comment (harshda15125) [reply src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs]: resolved

File: src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs (hunk 4 above)
Comment (harshda15125) [reply src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs]: Resolved

File: src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs (hunk 1 above)
Comment (harshda15125) [reply src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs]: Resolved

File: src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs (hunk 2 above)
Comment (harshda15125) [reply src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs]: Resolved

File: tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/WebhookScenarios.cs [hunk 5]
```diff
@@ -54,71 +54,71 @@ public async Task WhenValidEventInNewEncContentPublishedEventReceivedWithTokenHa
         }
 
         [Test, Order(1)]
-        //New Cell
-        [TestCase("ID3_1NewCellScenario.JSON", TestName = "WhenICallTheWebhookWithOneNewCellScenario_ThenWebhookReturns200Response")]
-        [TestCase("ID4_2NewCellScenario.JSON", TestName = "WhenICallTheWebhookWithTwoNewCellScenario_ThenWebhookReturns200Response")]
-        [TestCase("ID5_1NewCellWoNewAVCSUnit.JSON", TestName = "WhenICallTheWebhookWithOneNewCellScenarioWithourAVCSUoS_ThenWebhookReturns200Response")]
+        //New Cell
```
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/WebhookScenarios.cs:57]: Remove this white space

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 6]
```diff
@@ -46,11 +48,11 @@ public static async Task<bool> CheckXMLAttributes(JsonPayloadHelper jsonPayload,
             foreach (ZMAT_ACTIONITEMS item in result.IM_MATINFO.ACTIONITEMS)
             {
                 if (item.ACTION == "CREATE ENC CELL")
-                    Assert.That(VerifyCreateENCCell(item.CHILDCELL, item));
-                else if (item.ACTION == "CREATE AVCS UNIT OF SALE")
-                    Assert.That(VerifyCreateAVCSUnitOfSale(item.PRODUCTNAME, item));
-                else if (item.ACTION == "ASSIGN CELL TO AVCS UNIT OF SALE")
-                    Assert.That(VerifyAssignCellToAVCSUnitOfSale(item.CHILDCELL, item.PRODUCTNAME, item));
+                    Assert.That(VerifyCreateENCCell(item.CHILDCELL, item, correctionTag));
```
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:51]: Remove white spaces

**Review by Vishal14583 — CHANGES_REQUESTED** [review]
Few minor review comments for white spaces.

File: tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/WebhookScenarios.cs [hunk 7]
```diff
@@ -54,71 +54,71 @@ public async Task WhenValidEventInNewEncContentPublishedEventReceivedWithTokenHa
         }
 
         [Test, Order(1)]
-        //New Cell
-        [TestCase("ID3_1NewCellScenario.JSON", TestName = "WhenICallTheWebhookWithOneNewCellScenario_ThenWebhookReturns200Response")]
-        [TestCase("ID4_2NewCellScenario.JSON", TestName = "WhenICallTheWebhookWithTwoNewCellScenario_ThenWebhookReturns200Response")]
-        [TestCase("ID5_1NewCellWoNewAVCSUnit.JSON", TestName = "WhenICallTheWebhookWithOneNewCellScenarioWithourAVCSUoS_ThenWebhookReturns200Response")]
+        //New Cell        
+        [TestCase("ID3_1NewCellScenario.JSON", "Y", TestName = "WhenICallTheWebhookWithOneNewCellScenario_ThenWebhookReturns200Response")]
+        [TestCase("ID4_2NewCellScenario.JSON", "Y", TestName = "WhenICallTheWebhookWithTwoNewCellScenario_ThenWebhookReturns200Response")]
```
Comment (Sadha1501493) [inline tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/WebhookScenarios.cs:59]: as per TC ID 144369, correctionTag should be "N" for ID4_2NewCellScenario.JSON

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 8]
```diff
@@ -19,8 +19,10 @@ public class SAPXmlHelper
         private static JsonPayloadHelper UpdatedJsonPayload { get; set; }
         public static List<string> ListFromJson = new();
         public static List<string> ActionsListFromXml = new();
+        private static string weekNoTag = "202403";
```
Comment (Sadha1501493) [inline tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:22]: please declare fields read-only.

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 9]
```diff
@@ -455,6 +457,12 @@ private static bool VerifyAssignCellToAVCSUnitOfSale(string childCell, string pr
                             AttrNotMatched.Add(nameof(item.PRODUCT));
                         if (!item.PRODTYPE.Equals((GetProductInfo(unitOfSale.CompositionChanges.AddProducts)).ProductType))
                             AttrNotMatched.Add(nameof(item.PRODTYPE));
+                        if (!item.WEEKNO.Equals(weekNoTag))
```
Comment (Sadha1501493) [inline tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:460]: Please re-order the assertions or move the comment.

File: tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/WebhookScenarios.cs (hunk 7 above)
Comment (Pradeep15558) [reply tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/WebhookScenarios.cs:59]: Change the correction tag to N

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs (hunk 8 above)
Comment (Pradeep15558) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:22]: Declared the variable as read-only

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs (hunk 9 above)
Comment (Pradeep15558) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:460]: Move the comment section up.

File: tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/WebhookScenarios.cs (hunk 5 above)
Comment (Pradeep15558) [reply tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/WebhookScenarios.cs]: Removed the blank space

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs (hunk 6 above)
Comment (Pradeep15558) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs]: Removed the blank spaces

**Review by Vishal14583 — APPROVED** [review]
Looks good.

**Review by Sadha1501493 — APPROVED** [review]
LGTM

**Pull request merged**